src/github_runner_image_builder/cli.py

```diff
@@ -0,0 +1,209 @@
+# Copyright 2024 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""Main entrypoint for github-runner-image-builder cli application."""
+import argparse
+import itertools
+
+# Subprocess module is used to execute trusted commands
+import subprocess  # nosec: B404
+from pathlib import Path
+from typing import cast
+
+from github_runner_image_builder import builder, store
+from github_runner_image_builder.config import (
+    ACTION_INIT,
+    ACTION_LATEST_BUILD_ID,
+    ACTION_RUN,
+    IMAGE_OUTPUT_PATH,
+    LTS_IMAGE_VERSION_TAG_MAP,
+    ActionsNamespace,
+    BaseImage,
+    get_supported_arch,
+)
+
+
+def main(args: list[str] | None = None) -> None:
+    """Run entrypoint for Github runner image builder CLI.
+
+    Args:
+        args: Command line arguments.
+
+    Raises:
+        ValueError: If invalid action argument was supplied.
+    """
+    options = _parse_args(args)
+    if options.action == "init":
+        builder.initialize()
+        return
+    if options.action == "latest-build-id":
+        print(
+            store.get_latest_build_id(
+                cloud_name=options.cloud_name, image_name=options.image_name
+            ),
+            end="",
+        )
+        return
+    if options.action == "run":
+        _build_and_upload(
+            base=options.base,
+            cloud_name=options.cloud_name,
+            image_name=options.image_name,
+            keep_revisions=options.keep_revisions,
+            callback_script_path=options.callback_script_path,
+        )
+        return
+    raise ValueError("Invalid CLI action argument.")
+
+
+def _parse_args(args: list[str] | None = None) -> ActionsNamespace:
+    """Parse CLI arguments.
+
+    Args:
+        args: Command line arguments.
+
+    Returns:
+        An object with parsed user inputs.
+    """
+    parser = argparse.ArgumentParser(
+        prog="Github runner image builder CLI",
+        description="Builds github runner image and uploads it to openstack.",
+    )
+    subparsers = parser.add_subparsers(
+        title="actions",
+        description="command modes for Github runner image builder CLI.",
+        dest="action",
+        required=True,
+    )
+    subparsers.add_parser(ACTION_INIT)
+    get_latest_id_parser = subparsers.add_parser(
+        ACTION_LATEST_BUILD_ID, description="Fetch the latest ID of the built image."
+    )
+    run_parser = subparsers.add_parser(ACTION_RUN, description="Build the image.")
+    get_latest_id_parser.add_argument(
+        dest="cloud_name",
+        help=(
+            "The cloud to use from the clouds.yaml file. The CLI looks for clouds.yaml in paths "
+            "of the following order: current directory, ~/.config/openstack, /etc/openstack."
+        ),
+        type=_non_empty_string,
```

Comment:
I'm surprised there's no built-in way to check this in argparse

Reply:
Me too, this has many stackoverflow questions :D